File: 2024/2/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

func directionOf(firstItem int, secondItem int) string {
	if secondItem < firstItem {
		return "decreasing"
	} else {
		return "increasing"
	}
}

func reportSafe(report []string) bool {

	direction := ""

	for index, item := range report {
		if index == len(report)-1 {
			continue
		}

		currentItemI, _ := strconv.Atoi(item)
		nextItemI, _ := strconv.Atoi(report[index+1])

		if nextItemI == currentItemI {
			return false
		}

		currentDirection := directionOf(currentItemI, nextItemI)

		if direction == "" {
			direction = directionOf(currentItemI, nextItemI)
		} else if currentDirection != direction {
			return false
		}

		distance := math.Abs(float64(currentItemI) - float64(nextItemI))

		if distance < 1 || distance > 3 {
			return false
		}
	}

	return true
}

func main() {
	log.Println("yo.")

	file, err := os.Open("input.txt")

	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading file: %v\n", err)
	}

	sumPart1 := 0

	for scanner.Scan() {
		line := scanner.Text() // Get the current line as a string
		// Process the line
		fmt.Println(line) // Example: print each line
		parts := strings.Fields(line)

		if reportSafe(parts) {
			sumPart1 += 1
		}
	}

	log.Println("sum part 1 ", sumPart1)
}

File: 2024/2/main.ts
import * as fs from "fs";

function directionOf(firstItem: number, secondItem: number): string {
	if (secondItem < firstItem) {
		return "decreasing";
	} else {
		return "increasing";
	}
}

function reportSafe(report: string[]): boolean {
	let direction: string = "";

	for (let i = 0; i < report.length - 1; ++i) {
		if (i == report.length - 1) {
			continue;
		}

		const item = report[i];

		let currentItemI = parseInt(item);
		let nextItemI = parseInt(report[i + 1]);

		if (nextItemI == currentItemI) {
			return false;
		}

		let currentDirection = directionOf(currentItemI, nextItemI);

		if (direction == "") {
			direction = directionOf(currentItemI, nextItemI);
		} else if (currentDirection != direction) {
			return false;
		}

		const distance = Math.abs(currentItemI - nextItemI);

		if (distance < 1 || distance > 3) {
			return false;
		}
	}

	return true;
}

interface Row {
	report: string[];
}

const fileContent = fs.readFileSync("input.txt", "utf-8");

// Split the file into lines and process each line
const rows: Array<Row> = fileContent
	.split("\n") // Split by new lines
	.filter((line: string) => line.trim() !== "") // Remove empty lines
	.map((line: string) => {
		const report = line.split(" "); // Split by space

		return {
			report,
		};
	})
	.filter((item: Row) => {
		return reportSafe(item.report);
	});

const sumPart1 = rows.length;

console.log(`sum part 1: ${sumPart1}`);
